### PR TITLE
Make Seedvault compile in AOSP properly

### LINKED ISSUE #97: Build inline in AOSP
To avoid having to use prebuilt APKs

### PR DESCRIPTION
**Pull request by mikeNG**
* Fixes #97

### CODE REVIEW

**Comment by chirayudesai** [comment]
`./gradlew androidDependencies` prints the deps, can be used for verification.

File: app/build.gradle
Comment (chirayudesai) [inline app/build.gradle:125]: Move this to it's own commit

Comment (chirayudesai) [reply app/build.gradle:125]: Also add a comment at the top here (around dependencies {) to make sure to update libs in the libs dir / update libs/Android.bp or so on.

**Comment by chirayudesai** [comment]
@stevesoltys do we still want to keep the prebuilt stuff around?

IMO we can drop it in a separate commit, which also ensures that we'll always keep this AOSP build working.

Anyone can still use the prebuilts if they want, we just don't provide an easy way to do it since it's no longer needed.

**Comment by stevesoltys** [comment]
@chirayudesai

I'm away for the weekend so I won't be able to properly review this PR / your suggestion until Sunday, but as long as we still have CI builds working I think we can drop prebuilts without issue.

Though, if there is no additional development / maintenance effort required to keep the prebuilt repo, I don't see the harm in leaving it either. Some people may still find it useful, and they won't have to change their manifests again either.

I can give a more definitive response to that on Sunday, hopefully I'm not blocking anything in the meantime.

File: app/build.gradle
Comment (grote) [reply app/build.gradle:125]: Maybe also add a note about how to retrieve the prebuild aars, either from local gradle cache or maven repo. This will be handy when we want to bump this alpha version to something more stable.

**Comment by stevesoltys** [comment]
@chirayudesai 

I think that if we can't get CI builds working inside the AOSP source tree (without significant build server resources) it's preferable to continue with prebuilt binaries.

I say this because being able to compile against `android.jar` is the most lightweight way to validate that (at a minimum) pull requests are able to build in AOSP (as long as all the dependencies are there).

I feel there isn't much maintenance effort required to keep them either. Maybe we can consider removing them if we find that they are a burden to maintain? Let me know if you can think of anything that makes them problematic.

**Comment by chirayudesai** [comment]
Pending test confirmation before merge. Backup / restore / going through settings options should all work.

@stevesoltys Compiling against `android.jar` won't verify whether it builds in AOSP at all, as the issues here were all due to the build system differences.

I just wanted to make sure everybody used this build in AOSP instead of prebuilts to make sure it got the most testing and stayed updated / working. Prebuilts won't be a burden to maintain I guess, maybe we can add a warning / some info there about this - and mark this as the preferred way?

What I mean is something like: https://www.phoronix.com/scan.php?page=news_item&px=Torvalds-Rust-Kernel-K-Build

**Comment by grote** [comment]
I think we just keep the current CI setup, but I don't think it is related to keeping a prebuilt script around as the CI is doing the gradle build which doesn't say how building in AOSP will work, but at least it can catch other kinds of errors and also run our test cases.

**Comment by stevesoltys** [comment]
@chirayudesai It does validate that the code compiles against AOSP, as long as it's up to date. I'm not referring to actually building in the source tree, which is why I said "as long as the dependencies are there".

Yes, we can change the documentation to prefer this method over prebuilts.

**Comment by steadfasterX** [comment]
I guess that builds just in plain / pure AOSP right?

tried it on LineageOS  and got

```
error: packages/apps/Seedvault/Android.bp:48:18: unrecognized property "platform_apis"
error: packages/apps/Seedvault/Android.bp:57:1: unrecognized module type "prebuilt_etc"
error: packages/apps/Seedvault/Android.bp:64:1: unrecognized module type "prebuilt_etc"
```

**Comment by mikeNG** [comment]
> I guess that builds just in plain / pure AOSP right?
> 
> tried it on LineageOS (16.0) and got
> 
> ```
> error: packages/apps/Seedvault/Android.bp:48:18: unrecognized property "platform_apis"
> error: packages/apps/Seedvault/Android.bp:57:1: unrecognized module type "prebuilt_etc"
> error: packages/apps/Seedvault/Android.bp:64:1: unrecognized module type "prebuilt_etc"
> ```

It requires at least Android 10 base to compile properly. Works on AOSP and derivatives like CalyxOS or LineageOS 17.1

**Comment by relan** [comment]
I integrated this change into my custom build of AOSP 10 (not LineageOS derivative). Builds and works for me. Settings, backup, restore---no crashes.

**Comment by grote** [comment]
@relan awesome! Thanks for the feedback!

**Comment by chirayudesai** [comment]
Backup seems to work ok here, but restore from setupwizard via internal storage doesn't.

```
E AndroidRuntime: kotlin.UninitializedPropertyAccessException: lateinit property backView has not been initialized
E AndroidRuntime:        at com.stevesoltys.seedvault.ui.recoverycode.RecoveryCodeInputFragment.onActivityCreated(RecoveryCodeInputFragment.kt:74)
```

@relan what provider did you restore from, and through what way did you restore?
Additionally, do you have `WITH_DEXPREOPT` set for your AOSP build?

**Comment by chirayudesai** [comment]
Simple fix, seems to have worked.

File: app/src/main/java/com/stevesoltys/seedvault/ui/recoverycode/RecoveryCodeInputFragment.kt
Comment (chirayudesai) [inline app/src/main/java/com/stevesoltys/seedvault/ui/recoverycode/RecoveryCodeInputFragment.kt:51]: `backView = v.findViewById(R.id.backView)`

**Comment by relan** [comment]
> what provider did you restore from, and through what way did you restore?

That was internal storage. I launched the RestoreActivity manually (not through the Setup Wizard), it didn't ask for the 12-word password. I reproduced those steps and now I can see it didn't restore app data, only installed APKs. Manual reinstallation (via adb) of APKs does not trigger data restoration despite the Automatic restore option being on.

There are key-related exceptions in the log:
```
08-11 09:16:57.098 13406 13406 W KeyStore: KeyStore exception
08-11 09:16:57.098 13406 13406 W KeyStore: android.os.ServiceSpecificException:  (code 7)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.os.Parcel.createException(Parcel.java:2085)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.os.Parcel.readException(Parcel.java:2039)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.os.Parcel.readException(Parcel.java:1987)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.security.keystore.IKeystoreService$Stub$Proxy.get(IKeystoreService.java:978)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.security.KeyStore.get(KeyStore.java:236)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.security.KeyStore.get(KeyStore.java:225)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.security.keystore.AndroidKeyStoreSpi.engineGetCertificate(AndroidKeyStoreSpi.java:160)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at java.security.KeyStoreSpi.engineEntryInstanceOf(KeyStoreSpi.java:583)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at java.security.KeyStore.entryInstanceOf(KeyStore.java:1631)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at com.stevesoltys.seedvault.crypto.KeyManagerImpl.hasBackupKey(KeyManager.kt:57)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at com.stevesoltys.seedvault.ui.RequireProvisioningViewModel.recoveryCodeIsSet$name(RequireProvisioningViewModel.kt:23)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at com.stevesoltys.seedvault.settings.SettingsActivity.onStart(SettingsActivity.kt:44)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1432)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.app.Activity.performStart(Activity.java:7848)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.app.ActivityThread.handleStartActivity(ActivityThread.java:3294)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.app.servertransaction.TransactionExecutor.performLifecycleSequence(TransactionExecutor.java:221)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.app.servertransaction.TransactionExecutor.cycleToPath(TransactionExecutor.java:201)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:173)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.os.Handler.dispatchMessage(Handler.java:107)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.os.Looper.loop(Looper.java:214)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at android.app.ActivityThread.main(ActivityThread.java:7356)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at java.lang.reflect.Method.invoke(Native Method)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
08-11 09:16:57.098 13406 13406 W KeyStore: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

> Additionally, do you have WITH_DEXPREOPT set for your AOSP build?

I kept the default, so it should be on.

File: app/build.gradle
Comment (grote) [reply app/build.gradle:128]: [2.3.0-alpha06](https://developer.android.com/jetpack/androidx/releases/lifecycle#2.3.0-alpha06) is out with some bugfixes, maybe we should move to that directly?

**Comment by chirayudesai** [comment]
> > what provider did you restore from, and through what way did you restore?
> 
> That was internal storage. I launched the RestoreActivity manually (not through the Setup Wizard), it didn't ask for the 12-word password. I reproduced those steps and now I can see it didn't restore app data, only installed APKs. Manual reinstallation (via adb) of APKs does not trigger data restoration despite the Automatic restore option being on.
>

Ok, not asking for the password sounds like you didn't wipe, that's expected if so.
Not restoring app data is strange, here that works fine (I specifically changed some stuff like theme/wallpaper/to make it easier to verify.

We're on #seedvault on freenode if you want to discuss anything real-time btw.

**Comment by stevesoltys** [comment]
>@stevesoltys things like these are exactly why I wanted to get rid of prebuilts and just mandate this for inclusion with AOSP builds. That way it keeps getting tested more. The gradle build would already be used for development, and if somebody wanted they could still use gradle to compile an apk and ship that.

@chirayudesai I'm not understanding what you ran into / how prebuilts caused this. Could you explain please?

**Comment by chirayudesai** [comment]
My initial test with this had issues, but I retried afterwards and it seemed fine.

Scenarios to test:
1) Upgrade device to Android.bp, trying to make a new backup
2) New install, try to restore an old gradle backup.
3) New install, try to backup.

> > @stevesoltys things like these are exactly why I wanted to get rid of prebuilts and just mandate this for inclusion with AOSP builds. That way it keeps getting tested more. The gradle build would already be used for development, and if somebody wanted they could still use gradle to compile an apk and ship that.
> 
> @chirayudesai I'm not understanding what you ran into / how prebuilts caused this. Could you explain please?

Prebuilts didn't cause any issues, but Android.bp _may_ have.

My point was we should standardize on Android.bp to make sure it gets tested well all the time.

**Comment by stevesoltys** [comment]
@chirayudesai Ah, okay. We can prebuilts if you think that's preferable to ensure this new build method is well tested. I think I misunderstood before.

**Comment by mikeNG** [comment]
@chirayudesai I took a look and I don't see any other missing variable

**Comment by mikeNG** [comment]
I have tested the following a few times and I couldn't find any bugs caused by the Android.bp conversion:
1) make backup with unmodified build - backup-step-1
2) upgrade (no wipe) to Android.bp build
3) make another backup, backup-step-3
4) wipe
5) try to restore backup-step-1
6) wipe
7) try to restore backup-step-3
8) wipe
9) create a new backup, fresh device - backup-step-9

**Review by chirayudesai — APPROVED** [review]

**Review by grote — APPROVED** [review]

**Comment by chirayudesai** [comment]
Can merge this after a quick smoke test by @mikeNG 

It should compile now, I had github auto update - it merged, but that's fine since we're rebasing pull requests so we won't have double merge commits.

**Comment by mikeNG** [comment]
I have tested backup/wipe/restore and it still works as before. This can be merged.

**Pull request merged**
